phidgets_steppers/src/steppers_ros_i.cpp

```diff
@@ -0,0 +1,381 @@
+            steppers_->setRescaleFactor(channel, rescale_factor);
+
+            sprintf(str, "position_control_mode%02d", (int)channel);
+            bool position_control_mode = this->declare_parameter(str, true);
+            steppers_->setStepControlMode(channel, position_control_mode);
+
+            stepper_data_to_pub_[channel].stepper_sub =
+                std::make_unique<StepperSetter>(steppers_.get(), channel, this);
+
+            stepper_data_to_pub_[channel].last_position_val =
+                steppers_->getPosition(channel);
+            stepper_data_to_pub_[channel].last_velocity_val =
+                steppers_->getVelocity(channel);
+        }
+    } catch (const Phidget22Error& err)
+    {
+        RCLCPP_ERROR(get_logger(), "Steppers: %s", err.what());
+        throw;
+    }
+
+    stepper_pub_ = this->create_publisher<sensor_msgs::msg::JointState>(
+        "joint_states", 100);
+
+    if (publish_rate_ > 0.0)
+    {
+        double pub_msec = 1000.0 / publish_rate_;
+        timer_ = this->create_wall_timer(
+            std::chrono::milliseconds(static_cast<int64_t>(pub_msec)),
+            std::bind(&SteppersRosI::timerCallback, this));
+    } else
+    {
+        // If we are *not* publishing periodically, then we are event driven and
+        // will only publish when something changes (where "changes" is defined
+        // by the libphidget22 library).  In that case, make sure to publish
+        // once at the beginning to make sure there is *some* data.
+        publishLatestJointStates();
+    }
+}
+
+void SteppersRosI::publishLatestJointStates()
+{
+    sensor_msgs::msg::JointState msg;
+
+    msg.header.stamp = this->now();
+    msg.header.frame_id = frame_id_;
+
+    for (size_t channel = 0; channel < stepper_data_to_pub_.size(); channel++)
+    {
+        msg.position.push_back(stepper_data_to_pub_[channel].last_position_val);
+        msg.velocity.push_back(stepper_data_to_pub_[channel].last_velocity_val);
+    }
+    stepper_pub_->publish(msg);
+}
+
+void SteppersRosI::timerCallback()
+{
+    std::lock_guard<std::mutex> lock(stepper_mutex_);
+    publishLatestJointStates();
+}
+
+void SteppersRosI::positionChangeCallback(int channel, double position)
+{
+    if (static_cast<int>(stepper_data_to_pub_.size()) > channel)
+    {
+        std::lock_guard<std::mutex> lock(stepper_mutex_);
+        stepper_data_to_pub_[channel].last_position_val = position;
+
+        if (publish_rate_ <= 0.0)
+        {
+            publishLatestJointStates();
+        }
+    }
+}
+
+void SteppersRosI::velocityChangeCallback(int channel, double velocity)
+{
+    if (static_cast<int>(stepper_data_to_pub_.size()) > channel)
+    {
+        std::lock_guard<std::mutex> lock(stepper_mutex_);
+        stepper_data_to_pub_[channel].last_velocity_val = velocity;
+
+        if (publish_rate_ <= 0.0)
+        {
+            publishLatestJointStates();
+        }
+    }
+}
+
+void SteppersRosI::stoppedCallback(int channel)
```

Comment:
Yes I was trying to think of something useful I could do with a stopped callback and just left it there in case anyone had any ideas. At first I used it to publish 0.0 velocity on the JointState topic, but the changePosition callback already does that so that was redundant. I will remove it unless someone can think of something useful to do when the stepper stops.